### PR TITLE
Enable markdownlint rule (MD009)

### PR DESCRIPTION
**Pull request by Youssef1313**
Related to https://github.com/dotnet/runtime/pull/40884

cc: @jkotas @stephentoub there are too much violations of this rule in markdown files. To know if there are significant whitespaces or not, we will need to fix all the violations, and check if the diff shows any visual changes. Do you want me to fix that?

You can view the violations from [here](https://github.com/Youssef1313/runtime/runs/988600663?check_suite_focus=true).

### CODE REVIEW

File: .github/workflows/markdownlint.yml
Comment (Youssef1313) [inline .github/workflows/markdownlint.yml:1]: This is taken from dotnet/docs as is.

File: .github/workflows/markdownlint-problem-matcher.json
Comment (jkotas) [inline .github/workflows/markdownlint-problem-matcher.json:1]: Is this going to do anything in dotnet/runtime repo? We do not have github pipelines enable in this repo and we do not plan them enabled in foreseeable future (it is the long-term eventual goal).

Comment (Youssef1313) [reply .github/workflows/markdownlint-problem-matcher.json:1]: AFAIK, the problem matcher points to places of markdownlint errors in the "Files changed" tab in the PR.

Comment (Youssef1313) [reply .github/workflows/markdownlint-problem-matcher.json:1]: The markdownlint.yml file is the one responsible for running a GitHub action.

**Comment by jkotas** [comment]
Yes, fixing the end of line whitespaces in the markdown files is welcomed. I would do it in a separate PR from the infrastructure changes to enable markdown lint.

File: .github/workflows/markdownlint-problem-matcher.json
Comment (jkotas) [reply .github/workflows/markdownlint-problem-matcher.json:1]: Since we have Github actions disabled, does this manifest in the repo experience in any way?

Comment (Youssef1313) [reply .github/workflows/markdownlint-problem-matcher.json:1]: @jkotas I don't think it can be useful if GitHub actions are disabled.

**Comment by stephentoub** [comment]
@Youssef1313, is this PR still relevant at this point, or should it be closed?

**Comment by Youssef1313** [comment]
Should be closed I think given that the repo doesn't allow GitHub Actions.

**Comment by ViktorHofer** [comment]
@Youssef1313 We now use at least one github action in dotnet/runtime so feel free to reopen. We should be able to get this in.

**Comment by Youssef1313** [comment]
> 
> 
> @Youssef1313 We now use at least one github action in dotnet/runtime so feel free to reopen. We should be able to get this in.

That's great!

**Comment by Youssef1313** [comment]
@jkotas I'll delete that section and fix the new markdown violations.

File: docs/issues-pr-management.md
Comment (jkotas) [inline docs/issues-pr-management.md:32]:
```suggestion
```

**Review by jkotas — APPROVED** [review]

**Review by ViktorHofer — APPROVED** [review]
Fantastic, thanks a lot 👍

**Comment by danmosemsft** [comment]
Rebased. I will fix any remaining violations.

**Comment by danmosemsft** [comment]
OK rebased and fixed and now it's mergeable.

I see there's ~14,000 warnings other than MD009 and if I pass the auto fix flag it "fixes" about a third of them. If any of those other rules have value perhaps someone is interested in follow up PR's.

**Comment by danmosemsft** [comment]
Failure is https://github.com/dotnet/runtime/issues/45847

**Comment by danmosemsft** [comment]
summarizing discussion with @ViktorHofer - I removed the "push" section, so that this only runs on PR's, and does not trigger a Node install on rolling builds/CI. Our official builds, which produce shipping bits, do not run in Github pipelines, so this should not trigger Node there, regardless.

**Review by akoeplinger — APPROVED** [review]

**Comment by danmosemsft** [comment]
Thanks for the contribution @Youssef1313

**Comment by BruceForstall** [comment]
@danmosemsft Does this mean that changes to markdown files can lead to PR test failures? That seems extremely unnecessary.

**Comment by danmosemsft** [comment]
@BruceForstall I agree (my focus was on just getting this PR finished as I observed it was old). I think this step is low value, although perhaps more rules could be enabled like checking for broken links.

@safern is looking to see whether we can make this rule only trigger for markdown changes -- then if someone somehow checks in bad markdown, it won't affect 99% of PR's. Is that reasonable?

**Comment by Youssef1313** [comment]
> @safern is looking to see whether we can make this rule only trigger for markdown changes

This is already the case.
This workflow is only run when there are changes in md files.

**Comment by Youssef1313** [comment]
> @danmosemsft Does this mean that changes to markdown files can lead to PR test failures? That seems extremely unnecessary.

Only if the PR has bad markdown changea only.

**Comment by danmosemsft** [comment]
Also note that if it does fail on a markdown edit, it fails within moments of pushing to the PR, so you don't waste an hour or two waiting for the jobs to complete.

As mentioned, my care level about spacing is low, but there are more interesting rules (I have hand-fixed many broken links in the past and would rather not)

**Comment by BruceForstall** [comment]
> @safern is looking to see whether we can make this rule only trigger for markdown changes -- then if someone somehow checks in bad markdown, it won't affect 99% of PR's. Is that reasonable?

It would have no effect, since if you don't PR markdown and the rule doesn't fire, it won't have anything to complain about. Anything with markdown will be subject to the rule.

Your broken link example would be useful, but what if a link went away (even transiently) and my unrelated markdown change got flagged (well, assuming the rule applied to all markdown, not just the changed one). That would be annoying. And if it doesn't check all existing md for missing links, then there's no value anyway.

The question is whether we want to validate markdown at all. I hope we encourage markdown / documentation writing and don't discourage it in any way -- don't make it subject to the vagaries / flakiness of the CI system, for example, or make me wait for the CI to complete before merging.

One improvement would be if a markdown-only change only did markdown testing and nothing else.

**Comment by danmosemsft** [comment]
> It would have no effect, since if you don't PR markdown and the rule doesn't fire, it won't have anything to complain about. Anything with markdown will be subject to the rule.

The case I was concerned about was where a markdown error got in somehow and then breaks all PR jobs, who care nothing about markdown. It sounds like that won't happen.

> One improvement would be if a markdown-only change only did markdown testing and nothing else.

It does this already. I verified that yesterday. It's a rather nice experience - 2 mins from PR to green.

**Comment by safern** [comment]
Yep, the action is configured to only run when md files are changed:
https://github.com/dotnet/runtime/blob/118c0fd093dad025bb8ae37e85a77440e33dbb51/.github/workflows/markdownlint.yml#L4-L9

**Comment by danmoseley** [comment]
@Youssef1313 do you have any interest in following up to enable more useful rules, like for broken links, malformed tables and headings, etc?  I have spent my share of time fixing those things in the past. I would expect we could merge that PR much more quickly than this one as we already have the build step now.

**Comment by ViktorHofer** [comment]
@Youssef1313 thanks a lot for your contribution. +1, it would be fantastic to have more rules enabled.

**Comment by Youssef1313** [comment]
Sure, would be glad to help. May take a while though due to university exams starting next week :/

**Pull request merged**
